EIS-2819: Add Switch component

antoinerollindev opened the pull request:
EIS-2819: Add Switch component

- Added the Switch component, which is an alternative checkbox display
- Added a new story (based on the checkbox one)
- Added related tests
- Added Switch type in SelectionTypes enum

**EDIT**

- Added SelectionStatus.SWITCHED and SelectionStatus.UNSWITCHED
- Mixed status test removed
- Replaced the term 'checkbox' by 'switch' in the tests

-----

https://user-images.githubusercontent.com/66251236/114895786-11c94300-9e10-11eb-95fa-9966be6d9ece.mov

-----

- PR ui-toolkit-styles: [https://github.com/SymphonyPlatformSolutions/symphony-bdk-ui-toolkit-styles/pull/81](https://github.com/SymphonyPlatformSolutions/symphony-bdk-ui-toolkit-styles/pull/81)
- Jira ticket: [https://perzoinc.atlassian.net/browse/EIS-2819](https://perzoinc.atlassian.net/browse/EIS-2819)
- Doc: [https://perzoinc.atlassian.net/wiki/spaces/EIS/pages/1886097265/Switch+component](https://perzoinc.atlassian.net/wiki/spaces/EIS/pages/1886097265/Switch+component)

symphony-jean-michael commented on spec/components/selection/Switch.spec.tsx:
```diff
@@ -0,0 +1,86 @@
+
+    it('render with default props and initial value', () => {
+      const wrapper = shallow(
+        <Switch name="default-checkbox-name" value="default-checkbox-value" />
```
I suggest to rename to 'default-name' and 'default-value'

symphony-jean-michael commented on spec/components/selection/Switch.spec.tsx:
```diff
@@ -0,0 +1,86 @@
+      wrapper.unmount();
+    });
+
+    it('with "mixed" status', () => {
```
You can removed this test because it's for the 'mixed' state of a checkbox. (https://www.w3.org/TR/wai-aria-practices-1.1/examples/checkbox/checkbox-2/checkbox-2.html)
Not applicable for the Switch ;)

symphony-jean-michael commented on spec/components/selection/Switch.spec.tsx:
```diff
@@ -0,0 +1,86 @@
+        <Switch
+          name="checked-state-checkbox-name"
+          value="checked-state-checkbox-value"
+          status={SelectionStatus.CHECKED}
```
I suggest to use another status, maybe 'SelectionStatus.SWITCHED' ?

antoinerollindev replied:
Thank you for your review @symphony-jean-michael :)

Indeed, I forgot to adapt the test file. I removed the 'mixed status' test, and replaced the term 'checkbox' by 'switch'.

For the CHECKED status, if we use SWITCHED, we will have to adapt SelectionInput, and also adapt the classnames in the sass mixins.

Or we can avoid theses changes by having: 
```
enum SelectionStatus {
  CHECKED = 'checked',
  SWICHED = 'checked',
  MIXED = 'mixed',
  UNCHECKED = 'unchecked',
  UNSWITCHED = 'unchecked',
}
```

symphony-jean-michael commented:
Thank you @antoinerollindev for your contribution on UI-Toolkit! 😀 👍

antoinerollindev commented:
Thanks for your review @symphony-jean-michael 

- SelectionStatus.SWITCHED and SelectionStatus.UNSWITCHED added
- Mixed status test removed
- Replaced the term 'checkbox' by 'switch' in the tests

AliciaSymphony commented on src/components/selection/SelectionTypes.tsx:
```diff
@@ -1,6 +1,13 @@
 
+export const SelectionInputTypes = {
+  [SelectionTypes.CHECKBOX]: 'checkbox',
+  [SelectionTypes.SWITCH]: 'checkbox',
```
= 'switch' ?

antoinerollindev replied:
@AliciaSymphony 

This is the value that is injected in the `type` attribute of the `input`. `switch` is not a valid HTML input type.
It needs to be 'checkbox', or one of [https://www.w3schools.com/html/html_form_input_types.asp](https://www.w3schools.com/html/html_form_input_types.asp). 
<img width="282" alt="Screenshot 2021-04-16 at 10 36 24" src="https://user-images.githubusercontent.com/66251236/114997078-c65e7580-9e9f-11eb-8652-0eb65548eaf4.png">

AliciaSymphony replied:
alright perfect, I was not sure if it was a typo or on purpose :)

AliciaSymphony commented on spec/components/selection/Switch.spec.tsx:
```diff
@@ -0,0 +1,66 @@
+import { SelectionInput } from '../../../src/components/selection/SelectionInput';
+import SelectionStatus from '../../../src/components/selection/SelectionStatus';
+
+describe('Switch Component', () => {
```
Thanks for the tests!!
Even if Enzyme exists we use from now on only Testing Library

Of course don't rewrite them but it's a note for next time

symphony-adnane replied:
@antoinerollindev new tests are all written in Testing library could you please use it to test this component
Thank you !

AliciaSymphony reviewed (approved):
Good job!
